akka-http/src/main/scala/akka/http/scaladsl/server/directives/FramedEntityStreamingDirectives.scala

```diff
@@ -0,0 +1,245 @@
+   * By default the uploaded data is limited by the `akka.http.parsing.max-content-length`.
+   */
+  final def asSourceOfAsync[T](parallelism: Int, framing: FramingWithContentType)(implicit um: Unmarshaller[ByteString, T]): RequestToSourceUnmarshaller[T] =
+    asSourceOfAsync(parallelism)(um, framing)
+
+  /**
+   * Similar to [[asSourceOfAsync]], as it will apply at most `parallelism` unmarshallers in parallel.
+   *
+   * The source elements emitted preserve the order in which they are sent in the incoming [[HttpRequest]].
+   * If you want to sacrivice ordering in favour of (potential) slight performance improvements in reading the input
+   * you may want to use [[asSourceOfAsyncUnordered]] instead, which lifts the ordering guarantee.
+   *
+   * Refer to [[asSourceOf]] for more in depth-documentation and guidelines.
+   *
+   * See also [[MiscDirectives.withoutSizeLimit]] as you may want to allow streaming infinite streams of data in this route.
+   * By default the uploaded data is limited by the `akka.http.parsing.max-content-length`.
+   */
+  final def asSourceOfAsyncUnordered[T](parallelism: Int)(implicit um: Unmarshaller[ByteString, T], framing: FramingWithContentType): RequestToSourceUnmarshaller[T] =
+    asSourceOfInternal[T](framing, (ec, mat) ⇒ Flow[ByteString].mapAsyncUnordered(parallelism)(Unmarshal(_).to[T](um, ec, mat)))
+  /**
+   * Similar to [[asSourceOfAsync]], as it will apply at most `parallelism` unmarshallers in parallel.
+   *
+   * The source elements emitted preserve the order in which they are sent in the incoming [[HttpRequest]].
+   * If you want to sacrivice ordering in favour of (potential) slight performance improvements in reading the input
+   * you may want to use [[asSourceOfAsyncUnordered]] instead, which lifts the ordering guarantee.
+   *
+   * Refer to [[asSourceOf]] for more in depth-documentation and guidelines.
+   *
+   * See also [[MiscDirectives.withoutSizeLimit]] as you may want to allow streaming infinite streams of data in this route.
+   * By default the uploaded data is limited by the `akka.http.parsing.max-content-length`.
+   */
+  final def asSourceOfAsyncUnordered[T](parallelism: Int, framing: FramingWithContentType)(implicit um: Unmarshaller[ByteString, T]): RequestToSourceUnmarshaller[T] =
+    asSourceOfAsyncUnordered(parallelism)(um, framing)
+
+  // format: OFF
+  private final def asSourceOfInternal[T](framing: FramingWithContentType, marshalling: (ExecutionContext, Materializer) => Flow[ByteString, ByteString, NotUsed]#ReprMat[T, NotUsed]): RequestToSourceUnmarshaller[T] =
+    Unmarshaller.withMaterializer[HttpRequest, Source[T, NotUsed]] { implicit ec ⇒ implicit mat ⇒ req ⇒
+      val entity = req.entity
+      if (!framing.matches(entity.contentType)) {
+        val supportedContentTypes = framing.supported
+        FastFuture.failed(Unmarshaller.UnsupportedContentTypeException(supportedContentTypes))
+      } else {
+        val bytes = entity.dataBytes
```

Comment:
why does `dataBytes` have `Any` instead of `NotUsed` as mat val type? could use plain `via` here otherwise.